normits_demand/models/forecasting/EDGE_growth/growth_matrices.py

```diff
@@ -0,0 +1,149 @@
+# -*- coding: utf-8 -*-
+"""
+Module to get growth matrices ready for application to base matrices.
+
+Formats matrices, including converting to numpy arrays, and infills missing growth factors.
+"""
+# Built-Ins
+
+# Third Party
+import pandas as pd
+import numpy as np
+import caf.toolkit as ctk
+# Local Imports
+# pylint: disable=import-error,wrong-import-position
+# Local imports here
+from normits_demand.models.forecasting.edge_growth import utils
+
+# pylint: enable=import-error,wrong-import-position
+
+# # # CONSTANTS # # #
+
+# # # CLASSES # # #
+
+# # # FUNCTIONS # # #
+def prepare_growth_matrices(
+    demand_segments_df: pd.DataFrame,
+    factors_df: pd.DataFrame,
+    stations_lookup: pd.DataFrame,
+) -> dict:
+    """Prepare Growth Factor matrices into numpy matrices.
+
+    Function creates growth numpy matrix for all purposes and ticket types combination
+    on a station to station level.
+
+    Parameters
+    ----------
+    demand_segments_df : pd.DataFrame
+        models' demand segments information dataframe
+    factors_df : pd.DataFrame
+        EDGE growth factors dataframe
+    stations_lookup : pd.DataFrame
+        lookup for all model used stations zones and TLCs
+
+    Returns
+    -------
+    growth_matrices : dict
+        numpy growth matrices for all purposes and ticket types
+    """
+    # get list of purposes
+    purposes = demand_segments_df["Purpose"].unique()
+    # get list of ticket types
+    ticket_types = factors_df["TicketType"].unique()
+    # create a list of model used stations
+    # factors_df = utils.filter_stations(stations_lookup, factors_df)
+    # add stns zones
+    # merge on origin/production
+    factors_df = utils.merge_to_stations(
+        stations_lookup, factors_df, "ZoneCodeFrom", "ZoneCodeTo"
+    )
+    # keep needed columns
+    factors_df = factors_df[
+        [
+            "from_stn_zone_id",
+            "to_stn_zone_id",
+            "ZoneCodeFrom",
+            "ZoneCodeTo",
+            "purpose",
+            "TicketType",
+            "Demand",
+        ]
+    ]
+
+    # create matrices dictionary
+    growth_matrices = {i: {} for i in purposes}
+
+    # get growth matrices for each purpose/ticket type
+    for purpose in purposes:
+        for ticket_type in ticket_types:
+            mx_df = factors_df[
+                ["from_stn_zone_id", "to_stn_zone_id", "Demand"]
+            ].loc[
+                (factors_df["purpose"] == purpose)
+                & (factors_df["TicketType"] == ticket_type)
+            ]
+            # expand matrix
+            index = range(1, len(stations_lookup)+1)
+            growth_matrices[purpose][
+                ticket_type
+            ] = ctk.pandas_utils.long_to_wide_infill(mx_df, mx_df.columns[0], mx_df.columns[1], mx_df.columns[2], index, index, infill=0).values
```

Comment:
It's more efficient to build a separate dictionary for each ticket type inside the purpose loop, then assign that built dictionary to the purpose-keyed dictionary.

i.e. Rather than
```
growth_matrices = {i: {} for i in purposes}
for purpose in purposes:
    for ticket_type in ticket_types:
        growth_matrices[purpose][ticket_type] = some_operation()
```

Do this:
```
growth_matrices = dict()
for purpose in purposes:
    p_dict = dict()
    for ticket_type in ticket_types:
        p_dict[ticket_type] = some_operation()
    growth_matrices[purpose] = p_dict
```

In the second method you're having to do significantly less hashes to assign values to the dictionary